docs/source/cpp_tutorial.rst

```diff
@@ -0,0 +1,448 @@
+
+The following example demonstrates how to create `mdarray` matrices in both device and host memory, copy one to the other, and create mdspans out of them:
+
+.. code-block:: c++
+
+   #include <raft/core/device_mdarray.hpp>
+   #include <raft/core/host_mdarray.hpp>
+   #include <raft/core/copy.hpp>
+
+   raft::device_resources res;
+
+   int n_rows = 10;
+   int n_cols = 10;
+
+   auto device_matrix = raft::make_device_matrix<float>(res, n_rows, n_cols);
+   auto host_matrix = raft::make_host_matrix<float>(res, n_rows, n_cols);
+
+   // Set the diagonal to 1
+   for(int i = 0; i < n_rows; i++) {
+       host_matrix(i, i) = 1;
+   }
+
+   raft::copy(res, device_matrix.view(), host_matrix.view());
+
+Step 2: Generate some data
+==========================
+
+Let's build upon the fundamentals from the prior section and actually invoke some of RAFT's computational APIs on the device. A good starting point is data generation.
+
+.. code-block: c++
+
+   #include <raft/core/device_mdarray.hpp>
+   #include <raft/random/make_blobs.cuh>
+
+   raft::device_resources res;
+
+   int n_rows = 10000;
+   int n_cols = 10000;
+
+   auto dataset = raft::make_device_matrix<float, int64_t>(res, n_rows, n_cols);
+   auto labels = raft::make_device_vector<int64_t, int64_t>(res, n_rows);
+
+   raft::random::make_blobs(res, dataset.view(), labels.view());
+
+That's it. We've now generated a random 10kx10k matrix with points that cleanly separate into Gaussian clusters, along with a vector of cluster labels for each of the data points. Notice the `cuh` extension in the header file include for `make_blobs`. This signifies to us that this file contains CUDA device functions like kernel code so the CUDA compiler, `nvcc` is needed in order to compile any code that uses it. Generally, any source files that include headers with a `cuh` extension use the `.cu` extension instead of `.cpp`. The rule here is that `cpp` source files contain code which can be compiled with a C++ compiler like `g++` while `cu` files require the CUDA compiler.
+
+Since the `make_blobs` code generates the random dataset on the GPU device, we didn't need to do any host to device copies in this one. `make_blobs` is also asynchronous, so if we don't need to copy and use the data in host memory right away, we can continue calling RAFT functions with the `device_resources` instance and the data transformations will all be scheduled on the same stream.
+
+Step 3: Using brute-force indexes
+=================================
+
+Build brute-force index
+-----------------------
+
+Consider the `(10k, 10k)` shaped random matrix we generated in the previous step. We want to be able to find the k-nearest neighbors for all points of the matrix, or what we refer to as the all-neighbors graph, which means finding the neighbors of all data points within the same matrix.
+.. code-block:: c++
+
+   #include <raft/neighbors/brute_force.cuh>
```

Comment:
In this file modify RAFT to cuVS, in the code examples and in the documentation content.